### ISSUE TITLE
Intellisense confused matching forward declaration with type definition

### ISSUE THREAD

**Issue by zajo** [microsoft/vscode-cpptools#5803]
**Type: LanguageService**

```
Version: 1.47.2 (user setup)
Commit: 17299e413d5590b14ab0340ea477cdd86ff13daf
Date: 2020-07-15T18:22:06.216Z
Electron: 7.3.2
Chrome: 78.0.3904.130
Node.js: 12.8.1
V8: 7.8.279.23-electron.0
OS: Windows_NT x64 10.0.18363

C/C++ for Visual Studio Code
v0.29.0
```

In a header file I have:

```
class thread_local_signal_data
{
    ....
public:
    struct connection_list; weak_ptr<connection_list> cl_;
    ....
};
```

In a cpp file:

```
struct thread_local_signal_data::connection_list
{
  ....
};
```

Then, in the same cpp file, the following line:

```
shared_ptr<connection_list> cl=tlsd.cl_.lock();
```

For which Intellisense reports the following problem:

```
no suitable user-defined conversion from
"std::shared_ptr<boost::synapse::synapse_detail::thread_local_signal_data::connection_list>" to
"std::shared_ptr<boost::synapse::synapse_detail::thread_local_signal_data::connection_list>" exists
```

It appears that Intellisense fails to understand that the declaration and the definition of `thread_local_signal_data::connection_list` match (for brevity I omitted the namespaces in the snippets above).

For reference, here are direct links to the relevant lines in the GitHub repository:

* Forward declaration: https://github.com/zajo/synapse/blob/develop/include/boost/synapse/detail/common.hpp?ts=4#L72
* Definition: https://github.com/zajo/synapse/blob/develop/src/connect.cpp?ts=4#L140
* Incorrectly reported problem: https://github.com/zajo/synapse/blob/develop/src/connect.cpp?ts=4#L213

**Comment by sean-mcmanus** [comment]
Thanks a lot for reporting this. I've filed a bug on VS at https://developercommunity.visualstudio.com/content/problem/1122584/c-intellisense-confused-matching-forward-declarati.html (it's from a shared component).

**Comment by sean-mcmanus** [comment]
This was fixed with 1.4.1.

**Comment by zajo** [comment]
Thank you!
